azkaban-common/src/test/java/azkaban/executor/RunningExecutionsUpdaterTest.java

```diff
@@ -0,0 +1,153 @@
+  private RunningExecutionsUpdater updater;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    this.execution = new ExecutableFlow();
+    this.execution.setExecutionId(EXECUTION_ID_77);
+    this.activeExecutor = new Executor(1, "activeExecutor-1", 9999, true);
+    this.runningExecutions = new RunningExecutions();
+    this.runningExecutions.get().put(EXECUTION_ID_77, new Pair<>(
+        new ExecutionReference(EXECUTION_ID_77, this.activeExecutor), this.execution));
+    this.updater = new RunningExecutionsUpdater(this.updaterStage, this.alerterHolder,
+        this.commonMetrics, this.apiGateway, this.runningExecutions, this.executionFinalizer);
+    when(this.alerterHolder.get("email")).thenReturn(this.mailAlerter);
+  }
+
+  @After
+  public void tearDown() {
+    DateTimeUtils.setCurrentMillisSystem();
+  }
+
+  @Test
+  public void updateExecutionsStillRunning() throws Exception {
+    mockFlowStillRunning();
+    this.updater.updateExecutions();
+    verifyCallUpdateApi();
+    verifyZeroInteractions(this.executionFinalizer);
+  }
+
+  @Test
+  public void updateExecutionsExecutorDoesNotExist() throws Exception {
+    mockExecutorDoesNotExist();
+    this.updater.updateExecutions();
+    verifyFinalizeFlow();
+  }
+
+  @Test
+  public void updateExecutionsFlowDoesNotExist() throws Exception {
+    mockFlowDoesNotExist();
+    this.updater.updateExecutions();
+    verifyCallUpdateApi();
+    verifyFinalizeFlow();
+  }
+
+  @Test
+  public void updateExecutionsUpdateCallFails() throws Exception {
+    mockUpdateCallFails();
+    DateTimeUtils.setCurrentMillisFixed(System.currentTimeMillis());
+    for (int i = 0; i < this.updater.numErrorsBeforeUnresponsiveEmail; i++) {
+      this.updater.updateExecutions();
+      DateTimeUtils.setCurrentMillisFixed(
+          DateTimeUtils.currentTimeMillis() + this.updater.errorThreshold + 1L);
+    }
+    verify(this.mailAlerter).alertOnFailedUpdate(
+        this.activeExecutor, Collections.singletonList(this.execution), API_CALL_EXCEPTION);
+
+    // TODO change to checking if executor exist in the DB any more
+    verifyZeroInteractions(this.executionFinalizer);
+    // verify(this.executionFinalizer).finalizeFlow(this.execution, "TODO", null);
+  }
+
+  private void mockFlowStillRunning() throws Exception {
+    mockUpdateResponse();
+  }
+
+  private void mockExecutorDoesNotExist() {
+    this.runningExecutions.get().put(EXECUTION_ID_77, new Pair<>(
+        new ExecutionReference(EXECUTION_ID_77, null), this.execution));
+  }
+
+  private void mockUpdateCallFails() throws ExecutorManagerException {
+    doThrow(API_CALL_EXCEPTION).when(this.apiGateway).updateExecutions(any(), any());
+  }
+
+  private void verifyCallUpdateApi() throws ExecutorManagerException {
+    verify(this.apiGateway).updateExecutions(
+        this.activeExecutor, Collections.singletonList(this.execution));
+  }
+
+  private void mockFlowDoesNotExist() throws Exception {
```

Comment:
Do we need to add `("error", "Flow does not exist")` in the executionMap first before mocking the update response for this scenario?
Also, better to use `ConnectorParams.RESPONSE_ERROR` instead of `error`.

Reply:
> Do we need to add ("error", "Flow does not exist") in the executionMap first before mocking the update response for this scenario?

No, the order doesn't matter because it's the same map instance. I mean, the mock won't be called before the entry has been added to the map.

> Also, better to use ConnectorParams.RESPONSE_ERROR instead of error.

That's true.